### PR TITLE
Major refactor to allow lazy sync iterators.

### LINKED ISSUE #6: Optimization: Evaluate generators lazily
The current behavior eagerly loads the iterator returned by a generator function on its first call.

A more complex but more efficient solution would be to lazily evaluate the iterator from the parent function call as the wrapped function is called.

### PR DESCRIPTION
**Pull request by aebrahim**
Fixes #6.

Because the code was getting too unweildy and hard to follow, I refactored the iterator proxies to their own classes in a separate file.

### CODE REVIEW

File: once/_iterator_wrappers.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.generator: collections.abc.AsyncGenerator | None = func(*args, **kwargs)
+        self.finished = False
+        self.results: list = []
+        self.generating = False
+        self.lock = asyncio.Lock()
+
+    async def _yield_results(self) -> collections.abc.AsyncGenerator:
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:32]: I'm okay with the `_iterator_wrappers` module name to communicate that it's private, but within that, IMO, `yield_results` should have no prefix because it's the public interface of this class. That way we can also have real private methods that even __init__.py shouldn't call.

File: once/_iterator_wrappers.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        self.next_send = None
+
+    def _yield_results(self) -> collections.abc.Generator:
+        i = 0
+        # Fast path for subsequent calls will not require a lock
+        while True:
+            # If we on before the penultimate entry, we can return now. When yielding the last
+            # element of results, we need to be recording next_send, so that needs the lock.
+            if i < len(self.results) - 1:
+                yield self.results[i]
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:116]: This is limiting us to cPython by assuming there is a GIL that makes this list thread-safe to access. Is that a limitation you are okay with? Let's put a comment either way about the assumption.

File: once/_iterator_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            # i == len(self.results) - 1 or i == len(self.results)
+            # because the iterator could have moved in the interim. However, it will no longer
+            # move once self.finished.
+            if self.finished:
+                if i < len(self.results):
+                    yield self.results[i]
+                    i += 1
+                    continue
+                if i == len(self.results):
+                    return
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:129]: IMO this code is way too fancy trying to avoid grabbing the lock in the happy path. The lock is held super briefly since it's not held during `next()`, and python locks are super cheap, so this definitely isn't worth it. I spent like 30 minutes reading through python memory consistency models to figure out if this code had a bug, and I think I landed on in cPython, it happens to work at the moment, but is not guaranteed.

See https://peps.python.org/pep-0583/#uninitialized-values-flag

You're relying on a separate variable (`self.finished`) to act as a flag for a variable set in another thread (`self.results`) which is only true in a globally sequentially consistent memory model without synchronization. AFAIK (that PEP didn't really have a conclusion that I could see), Python does not guarantee such a model, nor does cPython, though it happens to likely have it at the moment.

I think we should lock liberally and have known bug-free code, and focus on true contention (`next()`) vs micro-optimizations before we have a need.

**Review by mattalbr — CHANGES_REQUESTED** [review]
More to review but this is the important stuff so far

File: once/_iterator_wrappers.py (hunk 1 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py]: done

File: once/_iterator_wrappers.py (hunk 3 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py]: I'm glad you brought this up - I took another crack at it and the code is way simpler now - PTAL!

File: once/_iterator_wrappers.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                        self.generating = True
+                else:
+                    action = _IteratorAction.YIELDING
+                    yield_value = self.results[i]
+            if action == _IteratorAction.WAITING:
+                continue
+            if action == _IteratorAction.YIELDING:
+                next_send = yield yield_value
+                i += 1
+                # If we have just sent the last element and we have not yet kicked off the next
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:151]: Why not just keep next_send as a local instead of sharing it across generators, like we do in async? Either way unlike in async we can't provide any guarantees to the caller about which send value gets used, but I think it keeps our code a bit simpler.

File: once/_iterator_wrappers.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                with self.lock:
+                    if i == len(self.results) and not self.generating:
+                        self.next_send = next_send
+                continue
+            if action == _IteratorAction.GENERATING:
+                assert self.generator is not None
+                try:
+                    result = self.generator.send(next_send)
+                except StopIteration:
+                    # This lock should be unnecessary, which by definition means there should be no
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:162]: I don't think it's unnecessary, since without it the order that these variables get set in other threads is undefined, so another thread could see for example `self.generating = False` `self.generator = None` and `self.finished = False` (python makes no guarantees about order without synchronization), which would result in trying to generate another result, which would fail our `assert self.generator is not None`

File: once/_iterator_wrappers.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+                async with self.lock:
+                    self.results.append(next_val)
+                    generating = None
+                    self.generating = None
+
+            send = yield next_val
+            i += 1
+
+
+class _IteratorAction(enum.Enum):
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:95]: I like this, maybe in a subsequent PR I'll use it for async too.

File: once/_iterator_wrappers.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                    result = self.generator.send(next_send)
+                except StopIteration:
+                    # This lock should be unnecessary, which by definition means there should be no
+                    # contention on it, so we use it to preserve our assumptions about variables which
+                    # are modified under lock.
+                    with self.lock:
+                        self.finished = True
+                        self.generator = None  # Allow this to be GCed.
+                        self.generating = False
+                else:
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:169]: nit: It's hard to tell if this `else` is on the `if` or the `try`. I think you could just in-line the else below `result = self.generator.send(next_send)`, like:

```python
try:
    result = self.generator.send(next_send)
    with self.lock:
        self.results.append(result)
        self.generating = False
except StopIteration:
    ...
```

**Review by mattalbr — CHANGES_REQUESTED** [review]
Will read through the tests in detail on the next pass

File: once/_iterator_wrappers.py (hunk 4 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py:151]: I just felt like this was easier to reason about and test for now. I'm happy to explore making a local variable in a subsequent PR

File: once/_iterator_wrappers.py (hunk 5 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py:162]: Only one thread at a time is allowed to have its state be `GENERATING`, so no other threads will ever access the code where `if action == _IteratorAction.GENERATING`.

File: once/_iterator_wrappers.py (hunk 7 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py:169]: is this any clearer?

File: once/_iterator_wrappers.py (hunk 5 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py:162]: removed comment per conversation

File: once/_iterator_wrappers.py (hunk 4 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py:151]: done

File: once/_iterator_wrappers.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                continue
+            assert self.generator is not None
+            try:
+                result = self.generator.send(next_send)
+            except StopIteration:
+                with self.lock:
+                    self.finished = True
+                    self.generator = None  # Allow this to be GCed.
+                    self.generating = False
+            else:
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:160]: Me personally I still prefer the `else` inside the `try` but it's just personal preference. If you like this, this is fine.

File: once/_iterator_wrappers.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                else:
+                    action = _IteratorAction.YIELDING
+                    yield_value = self.results[i]
+            if action == _IteratorAction.WAITING:
+                continue
+            if action == _IteratorAction.YIELDING:
+                next_send = yield yield_value
+                i += 1
+                continue
+            if action != _IteratorAction.GENERATING:
```
Comment (mattalbr) [inline once/_iterator_wrappers.py:150]: How is this possible? There are only three options and we already exhausted 2, right?

File: once_test.py [hunk 10]
```diff
@@ -77,13 +156,30 @@ def test_different_args_same_result(self):
...
+
+        @once.once
+        def yielding_iterator():
+            nonlocal counter
+            for _ in range(3):
+                yield counter.get_incremented()
 
-        self.assertEqual(list(yielding_iterator()), [0, 1, 2])
-        self.assertEqual(list(yielding_iterator()), [0, 1, 2])
+        with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
```
Comment (mattalbr) [inline once_test.py:179]: nit: might want to abstract max_workers as an env variable or flag, or at least a constant that's reused somewhere, so we can easily tweak it.

File: once_test.py [hunk 11]
```diff
@@ -360,6 +456,68 @@ def value():
...
+            for i in range(1, 32):
+                result.append(gen.send(i))
+            return result
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
+            results = list(executor.map(call_iterator, range(32)))
+        for result in results:
+            self.assertEqual(result, list(range(32)))
+
+    def test_receiving_iterator_parallel_execution_halting(self):
```
Comment (mattalbr) [inline once_test.py:500]: This test is non-obvious enough that a comment explaining it might be nice.

File: once_test.py [hunk 12]
```diff
@@ -360,6 +456,68 @@ def value():
...
+        @once.once
+        def receiving_iterator():
+            next = yield 0
+            while next is not None:
+                next = yield next
+
+        def call_iterator(n):
+            """Call the iterator but end early"""
+            gen = receiving_iterator()
```
Comment (mattalbr) [inline once_test.py:509]: What's the memory story for these generators that never get finished? When i was writing the async tests, I got warnings if I did that and had to make sure to close them all out.

**Review by mattalbr — CHANGES_REQUESTED** [review]
Nice! Last round I think

File: once/_iterator_wrappers.py (hunk 8 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py:160]: I prefer the explicitness - and we don't have another if to compete with for clarity here

File: once/_iterator_wrappers.py (hunk 9 above)
Comment (aebrahim) [reply once/_iterator_wrappers.py:150]: you're right, converted to assertion

File: once_test.py (hunk 10 above)
Comment (aebrahim) [reply once_test.py:179]: will do in upcoming cleanup PR.

File: once_test.py (hunk 11 above)
Comment (aebrahim) [reply once_test.py:500]: done

File: once_test.py (hunk 12 above)
Comment (aebrahim) [reply once_test.py:509]: AFAIK, it's safe in sync, unlike async where you are expected to await all of your promises.

File: once_test.py [hunk 13]
```diff
@@ -360,6 +456,70 @@ def value():
...
+        self.assertEqual(gen_2.send(None), 0)
+        self.assertEqual(gen_2.send(-1), 1)
+        self.assertEqual(gen_2.send(-1), 2)
+        self.assertEqual(gen_2.send(5), 5)
+        self.assertEqual(next(gen_2, None), None)
+        self.assertEqual(gen_1.send(None), 5)
+        self.assertEqual(next(gen_1, None), None)
+        self.assertEqual(list(receiving_iterator()), [0, 1, 2, 5])
+
+    def test_receiving_iterator_parallel_execution(self):
```
Comment (mattalbr) [inline once_test.py:480]: Oh I also think it would be good to have a sync analog to `test_iterator_lock_not_held_during_evaluation`

**Review by mattalbr — APPROVED** [review]
Woo we made it! Nice work!

File: once_test.py (hunk 13 above)
Comment (aebrahim) [reply once_test.py:480]: will do in upcoming refactor
